internal/download/download.go

```diff
@@ -62,21 +65,35 @@ func DownloadFile(url, outPath, header, token string) error {
 		time.Sleep(baseDelay*time.Duration(i+1) + jitter)
 	}
 
-	defer res.Body.Close()
-
+	var resErr error
 	switch res.StatusCode {
 	case 204:
 		fallthrough
 	case 404:
-		return fmt.Errorf("file not found")
+		resErr = fmt.Errorf("file not found")
 	case 401:
-		return fmt.Errorf("unauthorized, please ensure auth_header and auth_token inputs are set correctly and a valid token has been provided")
+		resErr = fmt.Errorf("unauthorized, please ensure auth_header and auth_token inputs are set correctly and a valid token has been provided")
 	default:
 		if res.StatusCode/100 != 2 {
-			return fmt.Errorf("failed to download file: %s", res.Status)
+			resErr = fmt.Errorf("failed to download file: %s", res.Status)
 		}
 	}
 
+	if resErr != nil {
+		defer res.Body.Close()
+		return nil, resErr
+	}
+
+	return res, nil
+}
+
+func DownloadFile(url, outPath, header, token string) error {
+	res, err := Fetch(url, header, token)
```

Comment:
We've extracted a bunch of logic from `DownloadFile` and placed it into `Fetch`. This function is then used to get a hold of the HTTP response for remote documents and use it to resolve an appropriate extension. `DownloadFile` does not give us the opportunity to do this because it expects callers to provide a path before knowing anything about the fetched content.